File: src/main/java/ru/ifmo/ted/security/XmlUserDetails.java
package ru.ifmo.ted.security;

import lombok.Getter;
import lombok.Setter;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import javax.xml.bind.annotation.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

@Setter
@XmlRootElement(name = "user")
@XmlAccessorType(XmlAccessType.FIELD)
class XmlUserDetails implements UserDetails {

    @Getter
    @XmlAttribute(name = "username")
    private String username = "empty";

    @Getter
    @XmlAttribute(name = "password")
    private String password = "none";

    @XmlElement(name = "authority")
    private List<String> authorities;

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        Collection<SimpleGrantedAuthority> simpleGrantedAuthorities = new ArrayList<>();
        for (String authority: authorities) {
            simpleGrantedAuthorities.add(new SimpleGrantedAuthority(authority));
        }
        return simpleGrantedAuthorities;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }

    @Override
    public String toString() {
        return "User " + username + ':' + password + ' ' + "[" + authorities + ']';
    }
}

File: src/main/kotlin/ru/ifmo/ted/security/XmlUserDetails.kt
package ru.ifmo.ted.security

import jakarta.xml.bind.annotation.*
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

@XmlRootElement(name = "user")
@XmlAccessorType(XmlAccessType.FIELD)
class XmlUserDetails(
    @XmlAttribute
    private val username: String,
    @XmlAttribute
    private val password: String,
    @XmlElement
    private val authorities: MutableCollection<XmlAuthority>
) : UserDetails {

    override fun getAuthorities() = authorities
    override fun getPassword() = password
    override fun getUsername() = username
    override fun isAccountNonExpired() = true
    override fun isAccountNonLocked() = true
    override fun isCredentialsNonExpired() = true
    override fun isEnabled() = true
}